File: mayonez-base/src/main/java/mayonez/physics/colliders/BulletBoxCollider.java
package mayonez.physics.colliders;

import mayonez.math.*;
import mayonez.math.shapes.*;
import mayonez.physics.dynamics.*;

/**
 * A box collider for fast moving objects that automatically extends itself along its path
 * to avoid tunneling.
 *
 * @author SlavSquatSuperstar
 */
public class BulletBoxCollider extends BoxCollider {

    private Vec2 displacement;
    private PhysicsBody rb;
    private boolean primaryAxisX; // Whether 0º is at +x or +y

    public BulletBoxCollider(Vec2 size) {
        super(size);
        primaryAxisX = true;
    }

    @Override
    protected void start() {
        super.start();
        displacement = new Vec2();
        rb = getPhysicsBody();
        if (rb == null) setEnabled(false);
    }

    @Override
    protected void update(float dt) {
        // Assume velocity direction is along object orientation
        // Assume no rotational velocity
        displacement.set(rb.getVelocity().mul(dt));
    }

    @Override
    public Shape getShape() {
        var addedSize = primaryAxisX
                ? new Vec2(displacement.len(), 0f)
                : new Vec2(0f, displacement.len());

        // New center is midpoint
        // New length is length + displacement
        return new Rectangle(
                transform.getPosition().add(displacement.mul(0.5f)),
                this.getSize().mul(transform.getScale()).add(addedSize),
                transform.getRotation()
        );
    }

//    @Override
//    protected void debugRender() {
//        var originalShape = super.getShape();
//        var nextShape = originalShape.translate(displacement);
//        var sweptShape = this.getShape();
//
//        getScene().getDebugDraw().drawShape(originalShape, Colors.BLUE);
//        getScene().getDebugDraw().drawShape(nextShape, Colors.RED);
//        getScene().getDebugDraw().drawShape(sweptShape, Colors.LIGHT_GREEN);
//    }

    // Setter Methods

    public void setPrimaryAxisX(boolean primaryAxisX) {
        this.primaryAxisX = primaryAxisX;
    }

}

File: mayonez-base/src/main/kotlin/mayonez/physics/colliders/BulletBoxCollider.kt
package mayonez.physics.colliders

import mayonez.math.*
import mayonez.math.shapes.*
import mayonez.physics.dynamics.*

/**
 * A box collider for fast moving objects that automatically extends itself along its path
 * to avoid tunneling.
 *
 * @author SlavSquatSuperstar
 */
class BulletBoxCollider(size: Vec2) : BoxCollider(size) {

    private lateinit var displacement: Vec2
    private var rb: PhysicsBody? = null

    // Collider Properties
    var sweepFactor: Float = 1f // How much to stretch swept box
    var isPrimaryAxisX: Boolean = true // Whether 0º is at +x or +y

    override fun start() {
        super.start()
        displacement = Vec2()
        rb = physicsBody
    }

    protected override fun update(dt: Float) {
        // Assume velocity direction is along object orientation
        // Assume no rotational velocity
        if (rb != null) {
            displacement.set(rb!!.velocity * dt * sweepFactor)
        }
    }

    override fun getShape(): Shape {
        val addedSize = if (this.isPrimaryAxisX) {
            Vec2(displacement.len(), 0f)
        } else {
            Vec2(0f, displacement.len())
        }

        // New center is midpoint
        // New length is length + displacement
        return Rectangle(
            transform.position + (displacement * 0.5f),
            (this.size * transform.scale) + addedSize,
            transform.rotation
        )
    }

//    protected override fun debugRender() {
//        val originalShape = super.getShape()
//        val nextShape = originalShape.translate(displacement)
//        val sweptShape = this.getShape()
//
//        scene.debugDraw.drawShape(originalShape, Colors.BLUE)
//        scene.debugDraw.drawShape(nextShape, Colors.RED)
//        scene.debugDraw.drawShape(sweptShape, Colors.LIGHT_GREEN)
//    }

}